File: src/main/java/com/healthpal/api/controller/UserController.java
package com.healthpal.api.controller;

import com.healthpal.api.model.User;
import com.healthpal.api.service.UserService;
import com.healthpal.api.service.AuthService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * UserController - REST API Controller for User Management
 * 
 * This controller is part of the MVC Layered Architecture (Controller/Presentation Layer)
 * It handles HTTP requests related to:
 * - User profile management
 * - User role assignment
 * - User activation/deactivation
 * - User data retrieval
 * 
 * Base URL: /api/users
 * 
 * @author HealthPal Team - Person A (Authentication + Admin + Core Architecture)
 * @version 1.0
 */
@Slf4j
@RestController
@RequestMapping("/api/users")
@CrossOrigin(origins = "*", maxAge = 3600)
public class UserController {

    @Autowired
    private UserService userService;

    @Autowired
    private AuthService authService;

    /**
     * Get current authenticated user profile
     * 
     * @return ResponseEntity with current user information
     */
    @GetMapping("/me")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<User> getCurrentUser() {
        log.info("Fetching current user profile");

        try {
            User currentUser = authService.getCurrentUser();
            return ResponseEntity.ok(currentUser);
        } catch (Exception e) {
            log.error("Error fetching current user: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    /**
     * Get user by ID
     * 
     * @param userId The ID of the user to fetch
     * @return ResponseEntity with user information
     */
    @GetMapping("/{userId}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<User> getUserById(@PathVariable Long userId) {
        log.info("Fetching user with ID: {}", userId);

        Optional<User> user = userService.getUserById(userId);
        return user.map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).build());
    }

    /**
     * Get user by username
     * 
     * @param username The username to search for
     * @return ResponseEntity with user information
     */
    @GetMapping("/username/{username}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<User> getUserByUsername(@PathVariable String username) {
        log.info("Fetching user with username: {}", username);

        Optional<User> user = userService.getUserByUsername(username);
        return user.map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).build());
    }

    /**
     * Get all active users (Admin only)
     * 
     * @return ResponseEntity with list of active users
     */
    @GetMapping("/active")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<List<User>> getAllActiveUsers() {
        log.info("Fetching all active users");

        List<User> users = userService.getAllActiveUsers();
        return ResponseEntity.ok(users);
    }

    /**
     * Get all admin users (Admin only)
     * 
     * @return ResponseEntity with list of admin users
     */
    @GetMapping("/admins")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<List<User>> getAllAdmins() {
        log.info("Fetching all admin users");

        List<User> admins = userService.getAllAdmins();
        return ResponseEntity.ok(admins);
    }

    /**
     * Get users by role (Admin only)
     * 
     * @param roleName The name of the role
     * @return ResponseEntity with list of users with the specified role
     */
    @GetMapping("/role/{roleName}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<List<User>> getUsersByRole(@PathVariable String roleName) {
        log.info("Fetching users with role: {}", roleName);

        List<User> users = userService.getUsersByRole("ROLE_" + roleName.toUpperCase());
        return ResponseEntity.ok(users);
    }

    /**
     * Update user profile
     * 
     * @param userId The ID of the user to update
     * @param user The updated user information
     * @return ResponseEntity with updated user information
     */
    @PutMapping("/{userId}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<User> updateUserProfile(@PathVariable Long userId, @RequestBody User user) {
        log.info("Updating user profile for user ID: {}", userId);

        try {
            User updatedUser = userService.updateUserProfile(userId, user);
            return ResponseEntity.ok(updatedUser);
        } catch (RuntimeException e) {
            log.error("Error updating user profile: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    /**
     * Change user password
     * 
     * @param userId The ID of the user
     * @param passwordChangeRequest Request containing old and new passwords
     * @return ResponseEntity with success message
     */
    @PostMapping("/{userId}/change-password")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, String>> changePassword(
            @PathVariable Long userId,
            @RequestBody Map<String, String> passwordChangeRequest) {

        log.info("Changing password for user ID: {}", userId);

        try {
            String oldPassword = passwordChangeRequest.get("oldPassword");
            String newPassword = passwordChangeRequest.get("newPassword");

            userService.changePassword(userId, oldPassword, newPassword);

            Map<String, String> response = new HashMap<>();
            response.put("message", "Password changed successfully");
            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            log.error("Error changing password: {}", e.getMessage());
            Map<String, String> response = new HashMap<>();
            response.put("error", e.getMessage());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }
    }

    /**
     * Assign role to user (Admin only)
     * 
     * @param userId The ID of the user
     * @param roleRequest Request containing role name
     * @return ResponseEntity with updated user information
     */
    @PostMapping("/{userId}/assign-role")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<User> assignRoleToUser(
            @PathVariable Long userId,
            @RequestBody Map<String, String> roleRequest) {

        log.info("Assigning role to user ID: {}", userId);

        try {
            String roleName = roleRequest.get("roleName");
            User updatedUser = userService.assignRoleToUser(userId, "ROLE_" + roleName.toUpperCase());
            return ResponseEntity.ok(updatedUser);
        } catch (RuntimeException e) {
            log.error("Error assigning role: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
        }
    }

    /**
     * Activate user account (Admin only)
     * 
     * @param userId The ID of the user to activate
     * @return ResponseEntity with updated user information
     */
    @PostMapping("/{userId}/activate")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<User> activateUser(@PathVariable Long userId) {
        log.info("Activating user ID: {}", userId);

        try {
            User activatedUser = userService.activateUser(userId);
            return ResponseEntity.ok(activatedUser);
        } catch (RuntimeException e) {
            log.error("Error activating user: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    /**
     * Deactivate user account (Admin only)
     * 
     * @param userId The ID of the user to deactivate
     * @return ResponseEntity with updated user information
     */
    @PostMapping("/{userId}/deactivate")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<User> deactivateUser(@PathVariable Long userId) {
        log.info("Deactivating user ID: {}", userId);

        try {
            User deactivatedUser = userService.deactivateUser(userId);
            return ResponseEntity.ok(deactivatedUser);
        } catch (RuntimeException e) {
            log.error("Error deactivating user: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    /**
     * Delete user (Admin only)
     * 
     * @param userId The ID of the user to delete
     * @return ResponseEntity with success message
     */
    @DeleteMapping("/{userId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Map<String, String>> deleteUser(@PathVariable Long userId) {
        log.info("Deleting user ID: {}", userId);

        try {
            userService.deleteUser(userId);
            Map<String, String> response = new HashMap<>();
            response.put("message", "User deleted successfully");
            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            log.error("Error deleting user: {}", e.getMessage());
            Map<String, String> response = new HashMap<>();
            response.put("error", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        }
    }
}

File: controller/userController.js
const UserService = require('../services/userService');
const AuthService = require('../services/authService');
const logger = require('../utils/logger');

class UserController {
    constructor() {
        this.userService = new UserService();
        this.authService = new AuthService();
    }

    async getCurrentUser(req, res) {
        logger.info('Fetching current user profile');
        try {
            const currentUser = await this.authService.getCurrentUser(req.user.id);
            res.status(200).json(currentUser);
        } catch (error) {
            logger.error(`Error fetching current user: ${error.message}`);
            res.status(404).json({ error: 'User not found' });
        }
    }

    async getUserById(req, res) {
        const { userId } = req.params;
        logger.info(`Fetching user with ID: ${userId}`);
        try {
            const user = await this.userService.getUserById(userId);
            if (!user) {
                return res.status(404).json({ error: 'User not found' });
            }
            res.status(200).json(user);
        } catch (error) {
            logger.error(`Error fetching user: ${error.message}`);
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    async getUserByUsername(req, res) {
        const { username } = req.params;
        logger.info(`Fetching user with username: ${username}`);
        try {
            const user = await this.userService.getUserByUsername(username);
            if (!user) {
                return res.status(404).json({ error: 'User not found' });
            }
            res.status(200).json(user);
        } catch (error) {
            logger.error(`Error fetching user: ${error.message}`);
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    async getAllActiveUsers(req, res) {
        logger.info('Fetching all active users');
        try {
            const users = await this.userService.getAllActiveUsers();
            res.status(200).json(users);
        } catch (error) {
            logger.error(`Error fetching users: ${error.message}`);
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    async getAllAdmins(req, res) {
        logger.info('Fetching all admin users');
        try {
            const admins = await this.userService.getAllAdmins();
            res.status(200).json(admins);
        } catch (error) {
            logger.error(`Error fetching admins: ${error.message}`);
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    async getUsersByRole(req, res) {
        const { roleName } = req.params;
        logger.info(`Fetching users with role: ${roleName}`);
        try {
            const users = await this.userService.getUsersByRole(`ROLE_${roleName.toUpperCase()}`);
            res.status(200).json(users);
        } catch (error) {
            logger.error(`Error fetching users by role: ${error.message}`);
            res.status(500).json({ error: 'Internal server error' });
        }
    }

    async updateUserProfile(req, res) {
        const { userId } = req.params;
        logger.info(`Updating user profile for user ID: ${userId}`);
        try {
            const updatedUser = await this.userService.updateUserProfile(userId, req.body);
            res.status(200).json(updatedUser);
        } catch (error) {
            logger.error(`Error updating user profile: ${error.message}`);
            res.status(404).json({ error: error.message });
        }
    }

    async changePassword(req, res) {
        const { userId } = req.params;
        logger.info(`Changing password for user ID: ${userId}`);
        try {
            const { oldPassword, newPassword } = req.body;
            await this.userService.changePassword(userId, oldPassword, newPassword);
            const response = { message: 'Password changed successfully' };
            res.status(200).json(response);
        } catch (error) {
            logger.error(`Error changing password: ${error.message}`);
            res.status(400).json({ error: error.message });
        }
    }

    async assignRoleToUser(req, res) {
        const { userId } = req.params;
        logger.info(`Assigning role to user ID: ${userId}`);
        try {
            const { roleName } = req.body;
            const updatedUser = await this.userService.assignRoleToUser(userId, `ROLE_${roleName.toUpperCase()}`);
            res.status(200).json(updatedUser);
        } catch (error) {
            logger.error(`Error assigning role: ${error.message}`);
            res.status(400).json({ error: error.message });
        }
    }

    async activateUser(req, res) {
        const { userId } = req.params;
        logger.info(`Activating user ID: ${userId}`);
        try {
            const activatedUser = await this.userService.activateUser(userId);
            res.status(200).json(activatedUser);
        } catch (error) {
            logger.error(`Error activating user: ${error.message}`);
            res.status(404).json({ error: error.message });
        }
    }

    async deactivateUser(req, res) {
        const { userId } = req.params;
        logger.info(`Deactivating user ID: ${userId}`);
        try {
            const deactivatedUser = await this.userService.deactivateUser(userId);
            res.status(200).json(deactivatedUser);
        } catch (error) {
            logger.error(`Error deactivating user: ${error.message}`);
            res.status(404).json({ error: error.message });
        }
    }

    async deleteUser(req, res) {
        const { userId } = req.params;
        logger.info(`Deleting user ID: ${userId}`);
        try {
            await this.userService.deleteUser(userId);
            const response = { message: 'User deleted successfully' };
            res.status(200).json(response);
        } catch (error) {
            logger.error(`Error deleting user: ${error.message}`);
            res.status(404).json({ error: error.message });
        }
    }
}

module.exports = UserController;